File: protocol_codec_test.ts
/**
 * TS test for protocol_codec.ts.
 *
 * Run with a TS runner:
 *   npx ts-node protocol_codec_test.ts
 * or
 *   npx tsx protocol_codec_test.ts
 */

import assert from "node:assert/strict"
import {
  MSG_AUDIO,
  MSG_CONTROL,
  MSG_TELEMETRY,
  FLAG_START,
  ENC_PCM16_LE,
  parsePacket,
  packAudioPacket,
  packTelemetryPacket,
  packControlPacket,
  AudioPayload,
  TelemetryPayload,
  ControlPayload,
} from "./protocol_codec"

function assertThrows(fn: () => void, message: string): void {
  let threw = false
  try {
    fn()
  } catch (err) {
    threw = true
    assert.ok(String(err).includes(message))
  }
  if (!threw) {
    throw new Error(`expected error containing: ${message}`)
  }
}

const zoneId = 3
const streamId = 11

const audio: AudioPayload = {
  encoding: ENC_PCM16_LE,
  channels: 1,
  sampleRateHz: 48_000,
  frameSamples: 480,
  audio: new Uint8Array(960),
}
const audioPacket = packAudioPacket(zoneId, streamId, 123, 480, audio, FLAG_START, true)
const audioDecoded = parsePacket(audioPacket)
assert.equal(audioDecoded.header.msgType, MSG_AUDIO)
assert.equal((audioDecoded.payload as AudioPayload).audio.length, 960)

const telemetry: TelemetryPayload = {
  status: 0x05,
  rssi: 99,
  ambientRms: 1000,
  bufferMs: 64,
  jitterMs: 5,
  packetLossPpm: 12,
  lastSeq: 123,
  streamId: 11,
}
const telemetryPacket = packTelemetryPacket(zoneId, streamId, 124, 960, telemetry)
const telemetryDecoded = parsePacket(telemetryPacket)
assert.equal(telemetryDecoded.header.msgType, MSG_TELEMETRY)
assert.equal((telemetryDecoded.payload as TelemetryPayload).lastSeq, 123)

const control: ControlPayload = {
  op: 1,
  zoneId,
  argU32: 0xDEADBEEF,
}
const controlPacket = packControlPacket(zoneId, streamId, 125, 1_440, control)
const controlDecoded = parsePacket(controlPacket)
assert.equal(controlDecoded.header.msgType, MSG_CONTROL)
assert.equal((controlDecoded.payload as ControlPayload).argU32, 0xDEADBEEF)

const audioPacketCorrupt = new Uint8Array(audioPacket)
// force one byte change while keeping crc enabled
audioPacketCorrupt[audioPacketCorrupt.length - 1] ^= 0x01
assertThrows(() => parsePacket(audioPacketCorrupt), "crc16 mismatch")

const badLen = new Uint8Array(audioPacket)
badLen[20] = 0
badLen[21] = 0
assertThrows(() => parsePacket(badLen), "payload_len mismatch")

console.log("protocol_codec_test.ts: checks complete")

File: protocol_codec_test.py
#!/usr/bin/env python3
"""
Basic regression tests for protocol_codec.py.
Run: python3 protocol_codec_test.py
"""

import secrets

from protocol_codec import (
    ControlPayload,
    Packet,
    MSG_AUDIO,
    MSG_CONTROL,
    MSG_TELEMETRY,
    FLAG_START,
    AudioPayload,
    TelemetryPayload,
    ENC_PCM16_LE,
    parse_packet,
    crc16_ccitt,
)


def make_audio_payload():
    audio = bytes(secrets.token_bytes(960))
    payload = AudioPayload(
        encoding=ENC_PCM16_LE,
        channels=1,
        sample_rate_hz=48000,
        frame_samples=480,
        audio=audio,
    )
    return payload


def run():
    aud = make_audio_payload()
    ap = Packet(MSG_AUDIO, zone_id=3, stream_id=11, seq=12345, timestamp=48000, flags=FLAG_START, payload=aud.pack())
    raw = ap.pack(include_crc=True)
    dec = parse_packet(raw)
    assert dec["header"].msg_type == MSG_AUDIO
    ap2 = dec["payload"]
    assert ap2.audio == aud.audio
    assert ap2.frame_samples == 480
    assert ap2.sample_rate_hz == 48000

    tlm = TelemetryPayload(
        status=0x05,
        rssi=99,
        ambient_rms=1000,
        buffer_ms=64,
        jitter_ms=5,
        packet_loss_ppm=12,
        last_seq=12345,
        stream_id=11,
    )
    tp = Packet(MSG_TELEMETRY, zone_id=3, stream_id=11, seq=54321, timestamp=96000, flags=0, payload=tlm.pack())
    traw = tp.pack(include_crc=False)
    tdec = parse_packet(traw)
    assert tdec["header"].msg_type == MSG_TELEMETRY
    tp2 = tdec["payload"]
    assert tp2.status == 0x05
    assert tp2.last_seq == 12345

    ctrl = ControlPayload(op=1, zone_id=3, arg_u32=0xDEADBEEF)
    cp = Packet(MSG_CONTROL, zone_id=3, stream_id=11, seq=999, payload=ctrl.pack(), timestamp=96000)
    craw = cp.pack(include_crc=False)
    cdec = parse_packet(craw)
    assert cdec["header"].msg_type == MSG_CONTROL
    cp2 = cdec["payload"]
    assert cp2.arg_u32 == 0xDEADBEEF

    # CRC rejection check
    bad = bytearray(raw)
    bad[-1] ^= 0x01
    try:
        parse_packet(bytes(bad))
        raise AssertionError("CRC corruption should fail")
    except ValueError as exc:
        assert "crc16 mismatch" in str(exc)

    # Corrupt payload length rejection check
    bad_len = bytearray(raw)
    bad_len[21] = 0x00
    bad_len[20] = 0x00
    try:
        parse_packet(bytes(bad_len))
        raise AssertionError("Payload length mismatch should fail")
    except ValueError as exc:
        assert "payload_len mismatch" in str(exc)

    print("protocol_codec_test.py: all checks passed")


if __name__ == "__main__":
    run()

